### ISSUE TITLE
Performance (on Windows)

### ISSUE THREAD

**Issue by BiBi-Abc** [Genivia/ugrep#70]
I tried out UGrep and benchmarked it against https://github.com/BurntSushi/ripgrep

I downloaded http://opus.nlpl.eu/download.php?f=OpenSubtitles/v2018/mono/OpenSubtitles.raw.en.gz

and downloaded UGrep for windows (2.5.6) and ripgrep 12.1.1. I EXTRACTED the zip file, and moved everything away so it was just UGrep and the text file. I then ran this:
```
ugrep -w "Sherlock [A-Z]\w+"
```

I cleared the standbylist (windows file cache)
It took about 19 seconds on Windows 64 Bit with a 24 thread CPU.

Then with ripgrep:
```
rg -w "Sherlock [A-Z]\w+"
```
It took on average 11 seconds.

I performed 5 tests.

**Comment by genivia-inc** [comment]
Sure, this is possible. There are different algorithmic details when matching regex. Sometimes one tool is faster than the other and vice versa.

Ugrep on Windows does not optimize file access at the moment, i.e. reading files in 16K chunks for the search window. We're also still improving the speed of ugrep by testing more platforms and OS to get the internal SIMD tradeoff formula better adjusted to a given machine. What does ugrep `-V` show?

**Comment by BiBi-Abc** [comment]
Hi! Thank you for your reply!

> 
> 
> What does ugrep `-V` show?

It shows
```
ugrep 2.5.6 WIN32 +avx2 +pcre2_jit +zlib +bzip2 +lzma +lz4
License BSD-3-Clause: <https://opensource.org/licenses/BSD-3-Clause>
Written by Robert van Engelen and others: <https://github.com/Genivia/ugrep>
```

**Comment by genivia-inc** [comment]
I'm curious how you cleared the Windows file cache without rebooting?

What does `ugrep --stats ...` show for the number of threads (worker jobs) in your run?

**Comment by genivia-inc** [comment]
A quick status update.

I’ve increased the initial search buffer window size to 64K. This shaved off about 30% of the search time of ugrep x64 on Windows for this large 13GB file. The performance of ugrep on MacOS is also a bit better (2.9 GHz Intel Core i7, 16 GB 2133 MHz LPDDR3) by about 15%.

However, this is not enough IMO. Searching large files (those at least above 1GB) should be better optimized in ugrep. The file read latencies and paging overhead accounts for most of the performance difference for very large files such as this 13GB file.

For the next phase of performance enhancements, we should look into prefetching on Windows, and also on Linux with `readahead` (`O_DIRECT`).

Smaller files (below 1GB) are read with mmap. This can be further optimized too, with `madvise MADV_SEQUENTIAL`.

**Comment by BiBi-Abc** [comment]
Hi! Thank you for your time in this project.


> 
> 
> I'm curious how you cleared the Windows file cache without rebooting?
> 
> What does `ugrep --stats ...` show for the number of threads (worker jobs) in your run?

I cleared windows file cache by using https://wj32.org/wp/software/empty-standby-list/

ugrep --stats shows
```
Searched 2 files with 16 threads in 1 directory: 1 matching
The following pathname selections and restrictions were applied:
  --no-hidden (default)
```

**Comment by genivia-inc** [comment]
> ugrep --stats shows
> ...
Are these the stats for the search on the `en.txt`? I'm just curious to check that the search is done on `en.txt` and only on `en.txt`.

**Comment by BiBi-Abc** [comment]
> 
> 
> > ugrep --stats shows
> > ...
> 
> Are these the stats for the search on the `en.txt`? I'm just curious to check that the search is done on `en.txt` and only on `en.txt`.

I redid the test searching only `en.txt`, instead of `en.txt` and `ugrep.exe`. Now --stats shows this:
```
Searched 1 file: 1 matching
The following pathname selections and restrictions were applied:
  --no-hidden (default)
```

**Comment by genivia-inc** [comment]
I've redesigned some of the ugrep internals to improve performance. The improvements makes ugrep run almost **TWICE** as fast on the 13GB file on my MacOS machine and on a high-performance Linux machine.

The main changes are:
- improved adaptive buffering and file reading/scanning speeds
- disabled mmap: mmap is slower than the improved file reading speed 
- files are opened with Windows CreateFile2 with `FILE_FLAG_SEQUENTIAL_SCAN`

Testing on an Intel i5 Windows machine shows significant improvements. However, I do not have access to a high-end Windows machine at the moment. Therefore, I cannot check if the improvements (with and without CreateFile2) yield any speedups, because I'm running into file read speed limitations. I created two versions, ugrep.exe (without CreateFile2) and ugrep_cf2.exe (with CreateFile2) downloadable from our secure web site:

- [ugrep.exe](https://www.genivia.com/files/ugrep.exe)
- [ugrep_cf2.exe](https://www.genivia.com/files/ugrep_cf2.exe)

Let me know if you are able to use these. But keep in mind that this is work in progress.

**Comment by BiBi-Abc** [comment]
Wow! That looks great! I tried it out, and it is much faster!

**Comment by genivia-inc** [comment]
> Wow! That looks great! I tried it out, and it is much faster!

Which one is faster? ugrep.exe or ugrep_cf2.exe or are both about the same?

**Comment by BiBi-Abc** [comment]
On a NVME SSD with gigabytes/s reads, ugrep_cf2 is very slightly faster. This also matches up with my previous testing, my .NET implementation of it used `FileStream`, which internally uses windows `CreateFile` api, and my C++ implementation with `fopen` was just slightly slower.

**Comment by BiBi-Abc** [comment]
I'm curious, how do you handle your multiline searching? For example, if the search could span across multiple lines, how would you read the file chunk by chunk?

Because what if the chunk that you match against is in the boundary of the regex?

**Comment by genivia-inc** [comment]
> I'm curious, how do you handle your multiline searching? Because what if the chunk that you match against is in the boundary of the regex?

It's a kind of magic... one golden glance of what should be.

**Comment by BiBi-Abc** [comment]
> > I'm curious, how do you handle your multiline searching? Because what if the chunk that you match against is in the boundary of the regex?
> 
> It's a kind of magic... one golden glance of what should be.

Hmm.

**Comment by BiBi-Abc** [comment]
> > I'm curious, how do you handle your multiline searching? Because what if the chunk that you match against is in the boundary of the regex?
> 
> It's a kind of magic... one golden glance of what should be.

What does this mean?

**Comment by genivia-inc** [comment]
Perhaps a bit tacky (sorry), but I'm just using a quote of the famous Queen song to point out that glancing at the solution is golden, while alluding to the (over-used) quote "_Any sufficiently advanced technology is indistinguishable from magic._" - Arthur C. Clarke.

One of the golden rules of high performance computing is to do as little work as possible and collect postponed work to perform more efficiently in a batch.

I've built my own [regex engine](https://github.com/Genivia/RE-flex) and expanded it to perform even better with ugrep. Also PCRE2 works much better with ugrep (option `-P`) than pcregrep itself. Pcregrep cannot handle large files and runs into buffering issues that require tinkering with its options to set the proper buffer sizes. The regex engines in ugrep use a form of "(hard) partial matching", e.g. [pcre2partial](http://pcre.org/current/doc/html/pcre2partial.html).

**Comment by genivia-inc** [comment]
I made a small mistake in ugrep.exe that I corrected and will release v3.0 soon.

**Comment by BiBi-Abc** [comment]
Wow! V3 looks great! I just discovered ugrep recently, but I'm going to use it more often now.
